flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/KnnModel.java

```diff
@@ -0,0 +1,378 @@
+            List<Integer> items = new ArrayList<>();
+            List<Double> metrics = new ArrayList<>();
+            while (!priorityQueue.isEmpty()) {
+                Tuple2<Double, Integer> result = priorityQueue.poll();
+                items.add(result.f1);
+                metrics.add(result.f0);
+            }
+            Collections.reverse(items);
+            Collections.reverse(metrics);
+            priorityQueue.clear();
+            return Tuple2.of(items, metrics);
+        }
+
+        /**
+         * @param input Input vector.
+         * @param topN Top N.
+         * @param priorityQueue Priority queue.
+         */
+        private void search(
+                DenseVector input,
+                Integer topN,
+                PriorityQueue<Tuple2<Double, Integer>> priorityQueue,
+                KnnModelData modelData) {
+            Tuple2<DenseVector, Double> sample = computeNorm(input);
+            Tuple2<Double, Integer> head = null;
+            for (int i = 0; i < modelData.getLength(); i++) {
+                List<Tuple2<Double, Integer>> values = computeDistance(sample, i);
+                for (Tuple2<Double, Integer> currentValue : values) {
+                    head = updateQueue(priorityQueue, topN, currentValue, head);
+                }
+            }
+        }
+
+        /**
+         * Updates queue.
+         *
+         * @param pq Queue.
+         * @param topN Top N.
+         * @param newValue New value.
+         * @param head Head value.
+         * @return Head value.
+         */
+        private <T> Tuple2<Double, T> updateQueue(
+                PriorityQueue<Tuple2<Double, T>> pq,
+                int topN,
+                Tuple2<Double, T> newValue,
+                Tuple2<Double, T> head) {
+            if (pq.size() < topN) {
+                pq.add(Tuple2.of(newValue.f0, newValue.f1));
+                head = pq.peek();
+            } else {
+                if (pq.comparator().compare(head, newValue) < 0) {
+                    Tuple2<Double, T> peek = pq.poll();
+                    assert peek != null;
+                    peek.f0 = newValue.f0;
+                    peek.f1 = newValue.f1;
+                    pq.add(peek);
+                    head = pq.peek();
+                }
+            }
+            return head;
+        }
+
+        /**
+         * Computes distance between sample and dictionary vectors.
+         *
+         * @param input Samples with l2 norm.
+         * @param index Dictionary vectors index.
+         * @return Distances.
+         */
+        private List<Tuple2<Double, Integer>> computeDistance(
+                Tuple2<DenseVector, Double> input, Integer index) {
+            Tuple3<DenseMatrix, DenseVector, int[]> data = modelData.getDictData().get(index);
+
+            DenseMatrix vectors = data.f0;
+            DenseMatrix distanceMatrix = new DenseMatrix(Objects.requireNonNull(vectors).numCols, 1);
```

Comment:
Since one of the dimension of `distanceMatrix` is 1, would it be more intuitive to make it a `DenseVector` instead of `DenseMatrix`?

Reply:
OK